File: c++/4Sum.cpp
/*
LeetCode 18 4Sum
Description : 

    Given an array nums of n integers, return an array of all the unique quadruplets [nums[a], nums[b], nums[c], nums[d]] such that:

    - 0 <= a, b, c, d < n
    - a, b, c, and d are distinct.
    - nums[a] + nums[b] + nums[c] + nums[d] == target

    You may return the answer in any order.

Question Link : https://leetcode.com/problems/4Sum/description/

Time: O(N^3)
Space: O(N)
*/

class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        // Create a set to store unique quadruplets
        set<vector<int>> result;
        
        // Sort the input array for easier processing
        sort(nums.begin(), nums.end());
        
        int n = nums.size();
        
        // Iterate through the array to fix the first two elements
        for(int i = 0; i < n - 3; i++) {
            for(int j = i + 1; j < n - 2; j++) {
                // Use two pointers to find the remaining elements
                int left = j + 1;
                int right = n - 1;
                long long newTarget = (long long)target - (long long)nums[i] - (long long)nums[j];
                
                while(left < right) {
                    long long sum = (long long)nums[left] + (long long)nums[right];
                    
                    if(sum == newTarget) {
                        // Found a valid quadruplet, add it to the result set
                        result.insert({nums[i], nums[j], nums[left], nums[right]});
                        left++;
                        right--;
                    } else if(sum < newTarget) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
        }
        
        // Convert the set to a vector for the final result
        vector<vector<int>> ans(result.begin(), result.end());
        
        return ans;
    }
};

File: JAVA/4Sum.java
//Problem link:https://leetcode.com/problems/4sum/


import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums); // Sort the input array in ascending order.
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        helper(nums, (long) target, 0, result, temp, 4); // Use long data type for target.
        return result; // Return the result list containing unique quadruplets.
    }

    private void helper(int[] nums, long target, int start, List<List<Integer>> result, List<Integer> temp, int numNeed) {
        if (numNeed != 2) {
            for (int i = start; i < nums.length - numNeed + 1; i++) {
                if (i > start && nums[i] == nums[i - 1]) {
                    continue; // Skip duplicates to avoid duplicate combinations.
                }
                temp.add(nums[i]); // Add the current number to the combination.
                helper(nums, target - nums[i], i + 1, result, temp, numNeed - 1); // Recursively find the next number(s).
                temp.remove(temp.size() - 1); // Remove the last number to backtrack.
            }
            return;
        }

        // If we need exactly 2 numbers, perform a two-pointer approach.
        int l = start;
        int r = nums.length - 1;
        while (l < r) {
            long total = (long) nums[l] + nums[r];
            if (total < target) {
                l++;
            } else if (total > target) {
                r--;
            } else {
                temp.add(nums[l]); // Add the left number to the combination.
                temp.add(nums[r]); // Add the right number to the combination.
                result.add(new ArrayList<>(temp)); // Store the valid quadruplet in the result list.
                temp.remove(temp.size() - 1); // Remove the right number to backtrack.
                temp.remove(temp.size() - 1); // Remove the left number to backtrack.
                l++;
                r--;
                while (l < r && nums[l] == nums[l - 1]) {
                    l++; // Skip duplicates on the left.
                }
            }
        }
    }
}
